Product purchase

K-Hir-pk18 opened the pull request:
購入時GIF
https://gyazo.com/6e1b12eb15575a5f3c675fb01453aac7

K-Hir-pk18 committed:
クレジット機能前

WHAT
・VIEWなど

WHY
・購入機能のINDEXなど

K-Hir-pk18 committed:
〜クレジット購入機能

WHAT
・１フォーム複数テーブル保存
・java PAYJP導入

WHY
・１オブジェクト1機能のため
・決済処理するため

K-Hir-pk18 committed:
商品購入機能

WHAT
・クレジットカードで
　商品を購入できる機能を実装

WHY
・アプリ上で購入希望者が
　目的の商品を購入できるため

ghost commented:
レビュー依頼ありがとうございます！
大変恐れ入りますが、実際の画面や機能ができているか確認させていただきたいので、下記キャプチャのご提出をお願いいたします。

```
- 購入に失敗してエラーが表示させる動画
- ログアウト状態のユーザーが商品購入ページに遷移しようとすると、ログインページに遷移する動画
- ログイン状態の出品者が、URLを直接入力して自身の出品した商品購入ページに遷移しようとすると、トップページに遷移する動画
- ログインユーザーがURLを直接入力して売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する動画
- 出品者・出品者以外にかかわらず、ログイン状態のユーザーが、URLを直接入力して売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画(編集機能要件より)
- ログアウト状態のユーザーが、URLを直接入力して売却済み商品の商品情報編集ページへ遷移しようとすると、ログインページに遷移する動画(編集機能要件より)
- ログイン状態の出品者でも、売却済みの商品に対しては「編集・削除ボタン」が表示されない動画（詳細機能要件より)
- テスト結果の画像
```

「Gyazo　使い方」とgoogleで検索し、ツールによって生成されたURLをこのPRに貼っていただければ幸いです。
その後、再度レビュー依頼フォームにてご提出をお願いいたします。

ghost commented:
また、プルリクエストにwhat,whyの記載がございません。
What（どのような実装をしているのか）と Why（なぜこの実装が必要なのか）を記載することで、
第三者にブランチでの作業内容を伝えることが可能でございます。
その後、再度レビュー依頼フォームにてご提出をお願いいたします。

K-Hir-pk18 commented:
kotonekuritam様

ご対応ありがとうございます。

商品購入機能に記載していたいのですが、隠れてわかりにくかったですね…。
以下ご確認お願いいたします。

WHAT
・クレジットカードで
　商品を購入できる機能を実装

WHY
・アプリ上で購入希望者が
　目的の商品を購入できるため

K-Hir-pk18 committed:
商品購入機能２

WHAT
・購入後商品の編集を不可に

WHY
・購入後の商品を編集可能にすると
　詐欺など犯罪に繋がりうるので

K-Hir-pk18 commented:
- 購入に失敗してエラーが表示させる動画
https://gyazo.com/222b7dfde2def7ed845b3cda7c8cd1d8

- ログアウト状態のユーザーが商品購入ページに遷移しようとすると、ログインページに遷移する動画
https://gyazo.com/2b32039b0eaf6c4bd2a5d5771bf89e5d

- ログイン状態の出品者が、URLを直接入力して自身の出品した商品購入ページに遷移しようとすると、トップページに遷移する動画
https://gyazo.com/376490216d44864c123321dd47896470

- ログインユーザーがURLを直接入力して売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する動画
https://gyazo.com/a08572bc78e14a7a77122dd25dfda80d

- 出品者・出品者以外にかかわらず、ログイン状態のユーザーが、URLを直接入力して売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画(編集機能要件より)
https://gyazo.com/fd5a66df084485ea11ba660fab612532

- ログアウト状態のユーザーが、URLを直接入力して売却済み商品の商品情報編集ページへ遷移しようとすると、ログインページに遷移する動画(編集機能要件より)
https://gyazo.com/a35d6c7c1308b6404cd282d99c1dcd82

- ログイン状態の出品者でも、売却済みの商品に対しては「編集・削除ボタン」が表示されない動画（詳細機能要件より)
https://gyazo.com/28d4e2dad3cc1f42d38685d47cc9013d

- テスト結果の画像
https://gyazo.com/9c440ef9887840fbd3874a80320163fb

shodai-ito commented on app/controllers/products_controller.rb:
```diff
@@ -25,6 +24,9 @@ def create
   def edit
+    if !@product.purchase_log.nil?
+      redirect_to root_path
+    end
```
こちらの条件はupdateとdestroyにも設定しましょう！
不正な更新や削除を防ぐ目的です！
メソッド化し、before_actionに記述することをお勧めします！

shodai-ito commented on app/controllers/purchases_controller.rb:
```diff
@@ -0,0 +1,31 @@
+    elsif current_user.id == @product.user_id
+      redirect_to root_path
+    elsif !@product.purchase_log.nil?
+      redirect_to root_path
```
こちらの条件はcreateにも設定しましょう！
また、
```
if user_signed_in? == false
      redirect_to user_session_path
```
こちらに関してはauthenticate_user!を使われた方がコードがシンプルになるかと存じます！

shodai-ito commented on app/controllers/purchases_controller.rb:
```diff
@@ -0,0 +1,31 @@
+
+  def create
+    @purchase_form = PurchaseForm.new(purchase_form_params)
+    @product = Product.find(params[:product_id])
```
`@product = Product.find(params[:product_id])`
こちらの記述が2回登場していますので、メソッド化してまとめましょう！

shodai-ito commented on app/models/purchase_form.rb:
```diff
@@ -0,0 +1,26 @@
+    validates :banch # , format: { with: /\A[ぁ-んァ-ン一-龥々]+\z/ }
+    validates :telephone, length: { in: 10..11 }
+    validates :telephone, format: { with: /\A[0-9]+\z/ }
+    validates :token
```
こちらuser_idとitem_idに対してもバリデーションを設定しましょう！
理由は、  今まではモデル間のアソシエーションを組むことで、自動的にpresence: trueが働いておりました。  ただ、フォームオブジェクトでは直接アソシエーションを組んでいないためにバリデーションが必要となります。

shodai-ito commented on app/models/purchase_form.rb:
```diff
@@ -0,0 +1,26 @@
+    validates :postal_code, format: { with: /\A\d{3}[-]\d{4}\z/ }
+    validates :prefecture_id, numericality: { other_than: 0, message: "can't be blank" }
+    validates :municipality # , format: { with: /\A[ぁ-んァ-ン一-龥々]+\z/ }
+    validates :banch # , format: { with: /\A[ぁ-んァ-ン一-龥々]+\z/ }
```
こちらのコメントアウトは不要であれば削除をお願いします！

shodai-ito commented on app/models/purchase_form.rb:
```diff
@@ -0,0 +1,26 @@
+  def save
+    @purchase_log = PurchaseLog.create(user_id: user_id, product_id: product_id)
+    Purchase.create(postal_code: postal_code, prefecture_id: prefecture_id, municipality: municipality, banch: banch, building_name: building_name, telephone: telephone, purchase_log_id: @purchase_log.id)
+    @product = Product.find(product_id)
+    Payjp.api_key = ENV['PAYJP_SECRET_KEY']
+    Payjp::Charge.create(
+      amount: @product.price,
+      card: token,
+      currency: 'jpy'
+    )
```
こちらの記述はcontrollerの方に記述いただくことが推奨されます。
移動しましょう！

K-Hir-pk18 commented:
ご担当者様

よろしくお願いいたします。

shodai-ito commented on app/views/products/index.html.erb:
```diff
@@ -132,9 +132,10 @@
+                  <%if product.purchase_log != nil %>
+                  <span>Sold Out!!</span>
+                  <% else %>
+                  <% end %>
```
この条件であればelseは不要です。
また、showのSoldOutの分岐と合わせるため、134行目の上から条件が分岐するよう記述しましょう！

shodai-ito commented on app/views/products/show.html.erb:
```diff
@@ -23,17 +23,18 @@
     </div>
 
-
+<%if @product.purchase_log != nil %>
+<% else %>
```
不要なelseはコードが見づらくなってしまいます。
<%if @product.purchase_log == nil %>にすればelseを使用しなくてもよくなりそうですね！
また、上記の条件は全てのボタン表示の有無についての分岐です。
28行目の<% if user_signed_in? %>と連結し、
`<% if user_signed_in? && @product.purchase_log == nil %>`
としてはいかがでしょうか？

shodai-ito commented on app/views/purchases/index.html.erb:
```diff
@@ -26,12 +28,16 @@
     <%# /支払額の表示 %>
 
-    <%= form_with id: 'charge-form', class: 'transaction-form-wrap',local: true do |f| %>
+<%#= form_with(url: product_purchase_path, local: true) do |f| %>
+  <%#= render '../shared/error_messages', model: @purchase %>
```
不要なコメントアウトは削除をお願いします！

shodai-ito commented on spec/models/purchase_form_spec.rb:
```diff
@@ -0,0 +1,78 @@
+  end
+
+  describe '商品購入', type: :model do
+    context '正常系 商品購入できる' do
+      it '必要な情報を適切に入力すると、商品の購入ができる' do
+        expect(@purchase_form).to be_valid
+      end
```
建物名がなくても購入できることもテストしましょう！

shodai-ito commented on spec/models/purchase_form_spec.rb:
```diff
@@ -0,0 +1,78 @@
+        @purchase_form.valid?
+        expect(@purchase_form.errors.full_messages).to include('Postal code is invalid')
+      end
+      it '配送先の情報として、都道府県が選択されていないとエラーが出ること' do
+        @purchase_form.prefecture_id = nil
+        @purchase_form.valid?
+        expect(@purchase_form.errors.full_messages).to include("Prefecture can't be blank")
+      end
```
都道府県が未選択の場合、無効なidが送信されるかと存じます。
nilでなく、無効なidを代入しましょう！

shodai-ito commented on spec/models/purchase_form_spec.rb:
```diff
@@ -0,0 +1,78 @@
+        @purchase_form.valid?
+        expect(@purchase_form.errors.full_messages).to include('Telephone is too long (maximum is 11 characters)')
+      end
+    end
```
こちらuser_idとitem_idに対してもテストを記述しましょう！

shodai-ito reviewed (commented):
レビュー依頼ありがとうございます！
コメントしましたので、ご確認をお願いいたします！

その後、再度レビュー依頼フォームのご提出をよろしくお願いいたします。

K-Hir-pk18 committed:
商品購入機能３

WHAT
・ 購入後編集の禁止
　→purchase_controller
　privateメソッド
　after_purchase_restriction作成

・controller purchase_controller
　→privateメソッドedit_judge
　→privateメソッドcreate_instance

・model purchase_form
　→validation追加
　　user_idとproduct_id

・model purchase_form
　→コメントアウトの削除

・view products/show 26,27行目
　→コード改良

・view urchase/index 36,37行目
　→コメントアウトの削除

・spec/test/purchase_form
　→正常系テスト追加（建物名）

　→異常系　都道府県エラー
　　代入をnilから0に変更

　→23行名変更
　　クレジット→商品購入できない

　→異常系　user_idとproduct_id追加

WHY
レビュー対応

・ 購入後編集の禁止
　不正編集の防止

・controller purchase_controller
　コードの改良、可読性向上

・model purchase_form
　validationの不備に対応

・model purchase_form
　可読性向上

・view products/show 26,27行目
　可読性向上

・view urchase/index 36,37行目
　可読性向上

・spec/test/purchase_form
　テストの抜け漏れに対応

　→異常系　都道府県エラー
　　正しい”空”の値に修正

　→23行名変更
　　内容の誤表記を修正

　→異常系　user_idとproduct_id追加
　　validationの抜け漏れに対応

K-Hir-pk18 commented:
ご担当者様

よろしくお願いいたします。

変更後テスト結果
https://gyazo.com/ab06f5fe95270ca4d1b0ae8cbd0776a0

chobi555 commented on app/controllers/purchases_controller.rb:
```diff
@@ -0,0 +1,36 @@
+  end
+
+  def edit_judge
+    if current_user.id == @product.user_id
+      redirect_to root_path
+    else !@product.purchase_log.nil?
+      redirect_to root_path
```
```ruby
if current_user.id == @product.user_id
      redirect_to root_path
    else !@product.purchase_log.nil?
      redirect_to root_path
```

こちらの条件分岐は論演算子を用いることでまとめることができますのでまとめてしまいましょう！

chobi555 commented on spec/models/purchase_form_spec.rb:
```diff
@@ -0,0 +1,92 @@
+        expect(@purchase_form.errors.full_messages).to include('Postal code is invalid')
+      end
+      it '配送先の情報として、都道府県が選択されていないとエラーが出ること' do
+        @purchase_form.prefecture_id = '0'
```
`@purchase_form.prefecture_id = '0'`
idですので、`'0'`の`''`は外しておきましょう！

K-Hir-pk18 replied:
chobi555様

ありがとうございます！

chobi555 reviewed (commented):
レビュー依頼ありがとうございます！
コメント致しました。
確認後、修正をお願い致します。

K-Hir-pk18 committed:
商品購入機能４

WHAT
・spec/models/purchase_form
　→都道府県コードの''除去

・contoroller purchases_controller
　→論理演算子に書換え

WHY
・spec/models/purchase_form
　→コードなので

・controllers/purchases_controller.rb
　→可読性の向上

K-Hir-pk18 commented:
ご担当射様

よろしくお願いいたします。

yogotaku commented on app/views/products/index.html.erb:
```diff
@@ -132,9 +132,10 @@
-                  <%# <% end %>
+                  <%if product.purchase_log != nil %>
+                  <span>Sold Out!!</span>
+                  <% else %>
```
```
<% else %>
```
こちらの`else`は不要なので削除しましょう。

yogotaku commented on spec/models/purchase_form_spec.rb:
```diff
@@ -0,0 +1,92 @@
+        @purchase_form.telephone = '012345678901'
+        @purchase_form.valid?
+        expect(@purchase_form.errors.full_messages).to include('Telephone is too long (maximum is 11 characters)')
+      end
```
電話番号について、以下のテストも追加しましょう。
```
電話番号が英数字混合では登録できないこと
```

K-Hir-pk18 replied:
yogotaku様

ありがとうございます！

K-Hir-pk18 replied:
ご担当者様

よろしくお願いいたします。


RSPEC
https://gyazo.com/f08cda48c756952807cd67e858b2c792

yogotaku reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認いたしました。
追加でコメントをしております。ご確認お願いします。

K-Hir-pk18 committed:
商品購入機能５

WHAT
・app/views/products/index.html.erb
　→137行目<%else%>の削除

・spec/models/purchase_form_spec.rb
　→電話番号　英字不可の追加
WHY
・app/views/products/index.html.erb
　→不要のため

・spec/models/purchase_form_spec.rb
　→正しい番号登録のため

yogotaku reviewed (commented):
レビュー依頼ありがとうございます。
修正確認いたしました。LGTMとさせていただきます！
ブランチをマージした後、次の実装へ進んでください。
お疲れ様でした！引き続き実装頑張っていきましょう！

K-Hir-pk18 commented:
yogotaku様

遅くまでのご対応ありがとうございます！
感謝です！！
